### ISSUE TITLE
Unexpected differences between tests sets provided for translation and tests sets generated with preprocess.py

### ISSUE THREAD

**Issue by tholiao** [pytorch/fairseq#592]
# Description
The provided test set (A) for the Gehring '17 model trained on WMT14 EN-DE has 2737 pairs of sentences, but the test set generated with fairseq-preprocess (B) has 3003 pairs of sentences. 

On inspection, it appears that every sentence which in (A) appears in (B), except that there are many more <UNK\> tokens in the source sentences in (B). 

# Details
I am running the prepare script then preprocessing the data as follows:
```
> bash prepare-wmt14en2de.sh --icml17
> TEXT=wmt14_en_de
> fairseq-preprocess --source-lang en --target-lang de \
--trainpref $TEXT/train --validpref $TEXT/valid --testpref $TEXT/test \
--destdir data-bin/wmt17_en_de --thresholdtgt 0 --thresholdsrc 0  \
--srcdict=data-bin/wmt14.en-de.fconv-py/dict.en.txt --tgtdict=data-bin/wmt14.en-de.fconv-pydict.de.txt
```

The source and target dictionaries are identical between the model, (A), and (B). 

I obtain a BLEU score of 25.7 on (A), but only 22.8 on (B)

# Examples
This sentence appears identically in both (A) and (B):
`16,000 joints of meat per week speak for themselves .`

This sentence appears in (B), but not in (A):
`100 years ago , Heinrich Hansjakob moved into his retirement home in Haslach , on Sunday his final years in the &quot; <unk> of &quot; were brought to life .`
Notably, this sentence appears in `newstest2014-deen-src.en.sgm`.

This sentence appears in both (A) and (B), but has <UNK\> tokens in (B):
`(A): 13.7 billion years ago , our cosmos was created by the Big Bang .`
`(B): 13.7 billion years ago , our <unk> s was created by the Big Bang .`

**Comment by myleott** [comment]
There are two wmt14 test sets with different numbers of examples. This can be seen with the sacrebleu tool:

```
> sacrebleu -t wmt14 -l en-de --echo src | wc -l
2737
> sacrebleu -t wmt14/full -l en-de --echo src | wc -l
3003
```

The wmt14 set (2737 examples) is the original evaluation data and wmt14/full is what people typically report scores on these days: https://github.com/mjpost/sacreBLEU/blob/master/sacrebleu.py#L275-L305

**Comment by myleott** [comment]
There shouldn't be such a big difference though. On other models I see differences of only ~0.1 BLEU going from wmt14 -> wmt14/full.

**Comment by tholiao** [comment]
Thanks for your reply. I'm able to get the correct numbers of sentences with wmt14/filtered. However, I'm still seeing <UNK> tokens in the dataset created with fairseq-preprocess.  Can you confirm that the following commands are correct?

```
> bash prepare-wmt14en2de.sh --icml17
> TEXT=wmt14_en_de
> fairseq-preprocess --source-lang en --target-lang de \
--trainpref $TEXT/train --validpref $TEXT/valid --testpref $TEXT/test \
--destdir data-bin/wmt17_en_de --thresholdtgt 0 --thresholdsrc 0  \
--srcdict=data-bin/wmt14.en-de.fconv-py/dict.en.txt --tgtdict=data-bin/wmt14.en-de.fconv-pydict.de.txt
```

**Comment by tholiao** [comment]
I'm now able to match the number of sentences.

I still see <UNK> tokens in the test set generated with fairseq-preprocess. Is this supposed to happen? #423 seems to indicate yes, but I am confused because the provided test set doesn't seem to have any. 

Can you confirm that the following steps are correct?
```
> cd examples/translation
> bash prepare-wmt14en2de.sh --icml17
> cd ../..
> TEXT=wmt14_en_de
> fairseq-preprocess --source-lang en --target-lang de \
--trainpref $TEXT/train --validpref $TEXT/valid --testpref $TEXT/test \
--destdir data-bin/wmt17_en_de --thresholdtgt 0 --thresholdsrc 0  \
--srcdict=data-bin/wmt14.en-de.fconv-py/dict.en.txt --tgtdict=data-bin/wmt14.en-de.fconv-pydict.de.txt
```
